### PR TITLE
Changes to support GANs

### LINKED ISSUE #535: Adding GAN support to TensorGraph
I think our current model with `TensorGraph` can't nicely support GANs. In order to train GANs, you need to train both a discriminator `D` and a generator `G`. The training of these two models is threaded together, with each updated for a few minibatches at a time. While updating the generator, the discriminator weights should be frozen. While updating the discriminator, the generator weights need to be frozen.

I have an idea for a new `TensorGraph` API to support this sort of training. We add a new `layers` kwarg to `fit`:

- `model.fit(feed_generator, layers=None)`

The idea is that `fit` can be instructed to train only some layers at a time. Note that there's now a semantic change:

- `Layer` is responsible for tracking the variables it introduces. 
  - `layer.variables` should be a list of all trainable variables created by that `Layer`.

@peastman and @lilleswing Does this basic design make sense? 

Hat tip to @enf for explaining GAN training to me in detail!

### PR DESCRIPTION
**Pull request by peastman**
See #535.  This turned out to be a bit more involved than I thought, but I think all the changes are reasonable.  For example, it now runs variable initializers in `build()` rather than `fit()`, since reinitializing them every time you call `fit()` obviously wouldn't make sense for this case.  Next step will be to create an example that trains a GAN.  That should give us a better sense of how this API works in practice.

There's one test failure, which is an actual bug that was exposed by these changes.  `WeaveLayer` is setting its `out_tensor` field to a list instead of a tensor.  Someone else needs to fix that.

### CODE REVIEW

**Comment by peastman** [comment]
The Python 3 build is failing with an unrelated error:

```
ERROR: Failure: ImportError (libicui18n.so.56: cannot open shared object file: No such file or directory)
```

Could this be related to the upgrade to Tensorflow 1.3?

**Comment by Dgelemi** [comment]
Which version of cuDNN are you using? Tensorflow 1.3 need version 6.
Not sure if it’s related though.


On 27 Sep 2017, at 19:49, peastman <notifications@github.com> wrote:

The Python 3 build is failing with an unrelated error:

ERROR: Failure: ImportError (libicui18n.so.56: cannot open shared
object file: No such file or directory)

Could this be related to the upgrade to Tensorflow 1.3?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/deepchem/deepchem/pull/855#issuecomment-332619525>, or mute
the thread

**Comment by peastman** [comment]
It's not using CUDA.  Travis doesn't have GPUs.

**Comment by lilleswing** [comment]
@peastman it was an issue importing rdkit not tensorflow -- no less confusing though.

**Comment by peastman** [comment]
Ok.  Are you fixing it?

**Comment by lilleswing** [comment]
I reran the build to see if it was reproducible -- if it is I can look at it at 3:00pm PST.

**Comment by peastman** [comment]
I already reran it once.  It happened both times.

**Comment by lilleswing** [comment]
Yeah looks like anaconda updated its version of icu and rdkit didn't pin it
https://anaconda.org/anaconda/icu
https://github.com/conda-forge/matplotlib-feedstock/issues/36

**Comment by peastman** [comment]
Great, thanks!

**Comment by proteneer** [comment]
pinging @greglandrum on this regarding what @lilleswing - should rdkit pin it?

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 1]
```diff
@@ -756,3 +838,40 @@ def __init__(self, tf_class, **kwargs):
 
   def __call__(self):
     return self.tf_class(**self.kwargs)
+
+
+class Submodel(object):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/tensor_graph.py]: Should a submodel be a top level class or perhaps inherit from one of our existing abstract classes?

**Comment by rbharath** [comment]
@peastman Would you mind raising a separate issue for the `WeaveLayer` failure that was exposed with a minimum failing example? We can fix that on a separate thread.

File: deepchem/models/tensorgraph/tensor_graph.py (hunk 1 above)
Comment (peastman) [reply deepchem/models/tensorgraph/tensor_graph.py]: Which one did you have in mind?

Comment (rbharath) [reply deepchem/models/tensorgraph/tensor_graph.py]: It's actually a little tricky now that I think about it. It feels like `Submodel` and `Model` should be related somehow but maybe that just doesn't work out under Python's current inheritance semantics.

Comment (peastman) [reply deepchem/models/tensorgraph/tensor_graph.py]: It's nothing to do with Python.  They're just different concepts.  Remember I suggested "submodel" should really be called "training objective" instead?  That's what it really is.  You preferred the name "submodel", so that's what I used, but remember that it *isn't* a model.  It's a training objective.

Comment (rbharath) [reply deepchem/models/tensorgraph/tensor_graph.py]: Let's rename it in that case :). I was envisioning something more related to a model object itself.

**Comment by greglandrum** [comment]
@proteneer, @lilleswing : we don't specify where ICU comes from and use whatever's on the system.

I can take a look at pinning a particular version based on the boost version we pin.

**Comment by rbharath** [comment]
@peastman Spent a bit more time thinking. I think we might simply be able to remove submodels/training-objectives and just allow a model to have multiple losses. Then a GAN is simply a model with two losses (discriminator loss and generator loss). Does this simplification work?

**Comment by peastman** [comment]
The submodel/objective can specify a list of layers, a loss, and an optimizer.  Just one of those isn't enough.  For example, the generator's loss function depends on the discriminator's variables, but it shouldn't modify those when optimizing the generator.

**Comment by rbharath** [comment]
@peastman This cgan tutorial looks amazing! Only thing I'd request is adding a paragraph describing how CGANs differ from regular GANs

**Comment by peastman** [comment]
> Only thing I'd request is adding a paragraph describing how CGANs differ from regular GANs

It already does: "A Conditional GAN (CGAN) allows additional inputs to the generator and discriminator that their output is conditioned on. For example, this might be a class label, and the GAN tries to learn how the data distribution varies between classes."

**Comment by peastman** [comment]
I believe this is now ready to merge.  `test_Weave_pickle()` is still failing, but that's a real error so it *should* be failing.  It will be fixed by #860.

**Comment by rbharath** [comment]
@peastman Everything looks to pass. Congrats on a major feature add!

LGTM
